osm2lanes/src/transform/tags_to_lanes/modes/bicycle.rs

```diff
@@ -48,105 +251,25 @@ pub(in crate::transform::tags_to_lanes) fn bicycle(
-        if tags.is(CYCLEWAY, "opposite") {
-            if !(road.oneway.into() && tags.is("oneway:bicycle", "no")) {
-                return Err(TagsToLanesMsg::unsupported_str(
-                    "cycleway=opposite without oneway=yes oneway:bicycle=no",
-                )
-                .into());
-            }
-            road.push_backward_outside(LaneBuilder::cycle_backward(locale));
-        }
-        // cycleway:FORWARD=*
-        if tags.is_cycleway(Some(locale.driving_side.into())) {
-            if tags.is(CYCLEWAY + locale.driving_side.tag() + "oneway", "no")
-                || tags.is("oneway:bicycle", "no")
-            {
-                road.push_forward_outside(LaneBuilder::cycle_both(locale));
-            } else {
-                road.push_forward_outside(LaneBuilder::cycle_forward(locale));
-            }
-        }
-        // cycleway:FORWARD=opposite_lane
-        if tags.is_any(
-            CYCLEWAY + locale.driving_side.tag(),
-            &["opposite_lane", "opposite_track"],
-        ) {
-            warnings.push(TagsToLanesMsg::deprecated_tags(
-                tags.subset(&[CYCLEWAY + locale.driving_side.tag()]),
-            ));
-            road.push_forward_outside(LaneBuilder::cycle_backward(locale));
-        }
-        // cycleway:BACKWARD=*
-        if tags.is_cycleway(Some(locale.driving_side.opposite().into())) {
-            if tags.is(
-                CYCLEWAY + locale.driving_side.opposite().tag() + "oneway",
-                "yes",
-            ) {
-                road.push_forward_inside(LaneBuilder::cycle_forward(locale));
-            } else if tags.is(
-                CYCLEWAY + locale.driving_side.opposite().tag() + "oneway",
-                "-1",
-            ) {
-                road.push_backward_outside(LaneBuilder::cycle_backward(locale));
-            } else if tags.is(
-                CYCLEWAY + locale.driving_side.opposite().tag() + "oneway",
-                "no",
-            ) || tags.is("oneway:bicycle", "no")
-            {
-                road.push_backward_outside(LaneBuilder::cycle_both(locale));
-            } else if road.oneway.into() {
-                // A oneway road with a cycleway on the wrong side
-                road.push_forward_inside(LaneBuilder::cycle_forward(locale));
-            } else {
-                // A contraflow bicycle lane
-                road.push_backward_outside(LaneBuilder::cycle_backward(locale));
-            }
-        }
-        // cycleway:BACKWARD=opposite_lane
-        if tags.is_any(
-            CYCLEWAY + locale.driving_side.opposite().tag(),
-            &["opposite_lane", "opposite_track"],
-        ) {
-            return Err(TagsToLanesMsg::unsupported_tags(
-                tags.subset(&[CYCLEWAY + locale.driving_side.opposite().tag()]),
-            )
-            .into());
-        }
+) -> Result<(), TagsToLanesMsg> {
+    let scheme = Scheme::from_tags(tags, locale, road.oneway, warnings)?;
```

Comment:
Separating tag parsing from modifying still looks like a good approach to me -- nice. A good test of the code maintainability will come when we start parsing tagged cycle lane widths or separator info.